src/number_theory/number_field/canonical_embedding.lean

```diff
@@ -0,0 +1,188 @@
+    fintype.card_subtype_compl, nat.add_sub_of_le (fintype.card_subtype_le _)],
+end
+
+instance [number_field K] : nontrivial (space K) :=
+begin
+  obtain ⟨w⟩ := infinite_place.nonempty K,
+  by_cases hw : is_real w,
+  { convert nontrivial_prod_left,
+    { convert @function.nontrivial _ _ _ real.nontrivial,
+      use ⟨w, hw⟩, },
+    exact nonempty_of_inhabited, },
+ { convert nontrivial_prod_right,
+   {  exact nonempty_of_inhabited, },
+   {  convert @function.nontrivial _ _ _ complex.nontrivial,
+      use ⟨w, not_is_real_iff_is_complex.mp hw⟩, }},
+end
+
+/-- The canonical embedding of a number field `K` of signature `(r₁, r₂)` into `ℝ^r₁ × ℂ^r₂`. -/
+def _root_.number_field.canonical_embedding : K →+* space K :=
+ring_hom.prod
+  (pi.ring_hom (λ w, w.prop.embedding))
+  (pi.ring_hom (λ w, w.val.embedding))
+
+lemma _root_.number_field.canonical_embedding_injective [number_field K] :
+  function.injective (number_field.canonical_embedding K) := ring_hom.injective _
+
+open number_field
+
+variable {K}
+
+@[simp]
+lemma apply_at_real_infinite_place (w : {w : infinite_place K // is_real w}) (x : K) :
+  (number_field.canonical_embedding K x).1 w = w.prop.embedding x :=
+by simp only [canonical_embedding, ring_hom.prod_apply, pi.ring_hom_apply]
+
+@[simp]
+lemma apply_at_complex_infinite_place (w : { w : infinite_place K // is_complex w}) (x : K) :
+  (number_field.canonical_embedding K x).2 w = embedding w.val x :=
+by simp only [canonical_embedding, ring_hom.prod_apply, pi.ring_hom_apply]
+
+lemma nnnorm_eq [number_field K] (x : K) :
+  ‖canonical_embedding K x‖₊ = finset.univ.sup (λ w : infinite_place K, ⟨w x, map_nonneg w x⟩) :=
+begin
+  rw [prod.nnnorm_def', pi.nnnorm_def, pi.nnnorm_def],
+  rw ( _ : finset.univ = {w : infinite_place K | is_real w}.to_finset
+    ∪ {w : infinite_place K | is_complex w}.to_finset),
+  { rw [finset.sup_union, sup_eq_max],
+    refine congr_arg2 _ _ _,
+    { convert (finset.univ.sup_map (function.embedding.subtype (λ w : infinite_place K, is_real w))
+        (λ w, (⟨w x, map_nonneg w x⟩ : nnreal))).symm using 2,
+      ext w,
+      simp only [apply_at_real_infinite_place, coe_nnnorm, real.norm_eq_abs,
+        function.embedding.coe_subtype, subtype.coe_mk, is_real.abs_embedding_apply], },
+    { convert (finset.univ.sup_map (function.embedding.subtype (λ w : infinite_place K,
+        is_complex w)) (λ w, (⟨w x, map_nonneg w x⟩ : nnreal))).symm using 2,
+      ext w,
+      simp only [apply_at_complex_infinite_place, subtype.val_eq_coe, coe_nnnorm,
+        complex.norm_eq_abs, function.embedding.coe_subtype, subtype.coe_mk, abs_embedding], }},
+  { ext w,
+    simp only [em (is_real w), set.mem_set_of_eq, finset.mem_union, set.mem_to_finset,
+      finset.mem_univ, ←infinite_place.not_is_real_iff_is_complex], },
+end
+
+lemma le_of_le [number_field K] (x : K) (r : ℝ) :
+  ‖canonical_embedding K x‖ ≤ r ↔ ∀ w : infinite_place K, w x ≤ r :=
```

Comment:
This doesn't strike me as a very informative name. Any better idea?

Reply:
`norm_le_iff`?

Reply:
Looks good, yes!